risc0/zkos/v1compat/src/kernel.s

```diff
@@ -0,0 +1,152 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+.equ USER_START_ADDR, 0x00010000
+.equ STACK_TOP, 0xfff00000
+.equ USER_REGS_ADDR, 0xffff0080
+.equ MEPC_ADDR, 0xffff0200
+.equ GLOBAL_OUTPUT_ADDR, 0xffff0240
+.equ ECALL_DISPATCH_ADDR, 0xffff1000
+.equ ECALL_TABLE_SIZE, 7
+.equ HOST_ECALL_TERMINATE, 0
+.equ HOST_ECALL_READ, 1
+.equ HOST_ECALL_SHA, 4
+.equ WORD_SIZE, 4
+.equ MAX_IO_BYTES, 1024
+.equ REG_T0, 5
+.equ REG_A0, 10
+.equ REG_A1, 11
+.equ REG_A2, 12
+.equ REG_A3, 13
+.equ REG_A4, 14
+
+.section .text
+.global _start
+_start:
+    # Load the global pointer from the place the linker put it
+    # Note: we definitely don't want this relaxed since the GP is invalid
+    .option push
+    .option norelax
+    la gp, __global_pointer$
+    .option pop
+
+    # Set the kernel stack pointer near the top of kernel memory
+    li sp, STACK_TOP
+
+    # Initialize the ecall dispatch table
+    li t0, ECALL_DISPATCH_ADDR
+    la t1, _ecall_dispatch
+    sw t1, 0(t0)
+
+    # Initialize useful constants
+    li t3, USER_REGS_ADDR
+    la t4, _ecall_table
+    li t5, ECALL_TABLE_SIZE
+    li t6, MAX_IO_BYTES
+
+    # Load the user program entry into MEPC
+    li a0, USER_START_ADDR
+    li a1, MEPC_ADDR
+    lw a2, 0(a0)
+    addi a2, a2, -WORD_SIZE
+    sw a2, 0(a1)
+
+    # Jump into userspace
+    mret
+
+_ecall_table:
+    j _ecall_halt
+    fence # input
+    j _ecall_software
+    j _ecall_sha
+    fence # bigint
+    fence # user
+    fence # bigint2
+
+_ecall_dispatch:
+    # load t0 from userspace
+    lw a0, REG_T0 * WORD_SIZE (t3)
+    # check that ecall request is within range
+    bge a0, t5, 1f
+    # adjust index so that it points to word sized entries
+    slli a0, a0, 2
+    # compute the table entry
+    add a1, t4, a0
+    # jump into dispatch table
+    jr a1
+1:
+    fence # panic
```

Comment:
trying to check my understanding here: this an atomic fence and is intended cause an illegal instruction error if we get to this point?